### ISSUE TITLE
Prevent user from exiting a route

### ISSUE THREAD

**Issue by Siyfion** [kadirahq/flow-router#318]
For example, say the user is halfway through typing a message in a blog post; if they click on a link that could take them away from the route, we need the ability to cancel the route change. Which we do, using `triggersExit` and `stop()`.

However, the reality is more complex; we don't want to stop the user from changing routes, only warn them that they will lose data if they do and give them a cancel option. So we popup a modal with a confirm / cancel option. The problem now is that this requires a callback, so the `triggersExit` function returns without having called `stop()` and thus continues, regardless of the user choice.

Could we not have a `continue()` function that *must* be called in order for the route to change, or alternatively a `callback(cancel)` type method structure?

**Comment by arunoda** [comment]
I think we can solve this problem very easily if you implement this logic on the click event. It's always better to keep the router as something lightweight.

**Comment by Siyfion** [comment]
@arunoda but the click event to what? **Every** link the the navigation bar, every link on the page, the "back" button in the browser...? That's a hell of a lot of click handling code.

**Comment by arunoda** [comment]
Ah okay. You mean, anything.

But we won't implementing anything async in the router layer. So, here you need to stop the route. Then when the user says okay, you need save the old URL and manually route it with `FlowRouter.go`.

If we add `continue()` like feature, triggers became kind a async. We don't like it. That's a core decision.

**Comment by Siyfion** [comment]
Okay, I can understand your reasoning for not wanting any async stuff, that's fair enough.. I'll give your idea a go in a MeteorPad.

**Comment by Siyfion** [comment]
@arunoda is there any way of getting the new route information when in a `triggersExit` function? The `context` variable seems to relate to the current route, not the one being transitioned too?

**Comment by arunoda** [comment]
Hmm. That's not possible. 
It's hard to get it in the current version. 
I can see that as a barrier for you. Let me think of a way.

**Comment by VladShcherbin** [comment]
@Siyfion When using the autoform package, it saves the data you inputed until it is saved to the database, so you can go from page to page and after you return - the data won't be lost. It even works on hot code reloads. You can also save the data to db, when it is entered, for example every 15 sec. 

I also don't think it is a router problem.

**Comment by Siyfion** [comment]
@VladShcherbin I disagree entirely; the data that could be lost isn't necessarily form data, in my particular example case, it's an SVG drawing. But I certainly don't want old drawings magically appearing when they go back into the drawing screen, without them having saved & loaded it first.

All I want is for the user to confirm that he/she is happy to lose any data they may have changed since they last saved.

The ***ONLY*** way I can detect a route change globally & reliably is within the router, thus it *is* a router issue. Where else can I handle this, other than attaching 60-or-so event handlers to every button, link, and heading that *could* be used to take the user to a different route?

**Comment by arunoda** [comment]
I'm labeling this as a question. We can't tell next route in the exit handler because of how we handle subscriptions and routes.

We are tracking it here: https://github.com/kadirahq/flow-router/issues/218

**Comment by Siyfion** [comment]
@arunoda Could you take a look at this: http://meteorpad.com/pad/Li7gESRXhuEJr745T/FlowRouter%20onExit%20stop()%20test

As I think it's a work-around for the above issue, ***but*** there is an issue. If you click navigate away once, then cancel. You can see that the URL has already changed. If you re-click navigate away, it just *does* it, seemingly never firing the exit trigger.

**Comment by awei01** [comment]
+1. It's a routing issue. Also, future implementation should probably watch `window.onunload` event as well.

**Comment by space-alien** [comment]
+1

**Comment by space-alien** [comment]
I have hacked together the following workaround for this problem which deals with the unwanted URL change.

It copes with the user pressing the back button, and also plays nicely with `window.onbeforeunload` (which you might want to use to catch navigating away from the app).

Swap out the Session variable with a check for unsaved changes, as needed.

I'd be interested to hear your comments.

```
var previousContext = {},
  requestedContext,
  isReverting;

FlowRouter.triggers.exit([function (context, redirect, stop) {
  // Before we leave, cache the current context.
  previousContext = context;
}]);

FlowRouter.triggers.enter([function (context, redirect, stop) {
  if (isReverting) {
    isReverting = false;

    // This route change is merely undoing the previous (postponed) route change.
    // We don't want to actually fire any route actions this time.
    stop();
  }
  else if (Session.get('unsavedChanges') && previousContext.path !== context.path) {
    // Cache the requested route until we actually have user confirmation.
    requestedContext = context;

    // It would have been preferable to prevent the current route change entirely, perhaps in triggers.exit().
    // But since we can't, we will 'undo' it manually instead.

    // Prevent the requested action from firing.
    stop();

    // Allow the current route handling to finish, otherwise we'd be left at the wrong URL.
    setTimeout(function () {
      // Go back to the original route
      FlowRouter.withReplaceState(function () {
        isReverting = true;
        FlowRouter.go(previousContext.route.pathDef, previousContext.params, previousContext.queryParams);

        // At last, it's safe to prompt the user and decide whether we really want to change route.
        if (window.confirm('You have unsaved changes. Are you sure you want to lose them?')) {
          Session.set('unsavedChanges', false);
          FlowRouter.go(requestedContext.route.pathDef, requestedContext.params, requestedContext.queryParams);
        }
      });
    }, 0);
  }
}]);
```

One slight drawback I don't think we can avoid is that if the user cancels, an extra history entry has been inserted. But we've taken care to ensure that this extra history entry will match the previous route exactly. Therefore, if the user does navigate back, the extra history entry shouldn't actually cause anything to happen - because FlowRouter is idempotent.

**Comment by space-alien** [comment]
I have reworked the solution in my previous comment and added more notes.

**Comment by marbemac** [comment]
We are trying to do something similar - prevent route change, and show a confirmation, if the current page entity has been modified but not saved. 

We expected stop() in an exit trigger to prevent the URL from changing, but it does not. Is this the intended behavior?

We'd like to avoid using a custom Link component for all of the anchor tags in the project, just to handle this case (as is suggested above). In fact, doing a e.preventDefault / e.stopPropogation in the anchor's onClick event still does not seem to prevent a route change.

**Comment by space-alien** [comment]
For anyone tracking this, I have revamped my [solution above](https://github.com/kadirahq/flow-router/issues/318#issuecomment-161678673). It now copes better with the user pressing the back button when there are unsaved changes, and the browser history is not polluted.

**Comment by szchenghuang** [comment]
@space-alien The remedy you proposed almost achieves all the expected results to me, except that there is still an entry created for the next route in the history.

Putting it in another way, the remedy rewinds from the next route, but leaves an unflattering trace.

**Comment by szchenghuang** [comment]
I tried many fixes and it turns out that the best remedy I worked out so far requires to dive deeper into the underlying `page.js` package. I raise a confirmation box in the `click` and `popstate` event listeners.

I have the `click` event listener fires as usual if the box is confirmed. Otherwise, have it return before displaying the associated page. Simple.

The `popstate` event listener requires a bit more challenge. First I add some logic to enable `page.js` to detect if the event is triggered with a forward or backward browsing, because I have to call `history.back()` or `history.forward()` respectively to rewind the UR. Second, as in the case of `click` event,  I again have this listener return before displaying the page associated to the next route if the confirmation is agreed.

**Comment by akvadrako** [comment]
This works for me:
```js
function wrapRouting(original) {
    return (...args) => {
        if(PageHelper.isRoutingAllowed())
            return original.apply(null, args);
    }
}

FlowRouter._page.redirect = wrapRouting(FlowRouter._page.redirect);
FlowRouter._page.show = wrapRouting(FlowRouter._page.show);
```

**Comment by macrozone** [comment]
@akvadrako : this is a easy solution and works good.

The only thing is, that it does not work, when user navigates back with browser. Any idea how to solve that?

**Comment by akvadrako** [comment]
Hello Marco,

Thanks - you should look at window.onbeforeunload for dealing with this at the browser level.

~ devin (via mobile)

On 28 Sep 2016, at 15:03, Marco Wettstein <notifications@github.com<mailto:notifications@github.com>> wrote:


@akvadrako<https://github.com/akvadrako> : this is a easy solution and works good.

The only thing is, that it does not work, when user navigates back with browser. Any idea how to solve that?

-

**Comment by NTruchaud** [comment]
Up here, 
Do we have an "official way" to do this now ?

**Comment by james1293** [comment]
@NTruchaud : I am also interested in an "official way", but [space-alien's method](https://github.com/kadirahq/flow-router/issues/318#issuecomment-161678673) seems to work quite well in the mean time.

@akvadrako : FYI - for me, `window.onbeforeunload` doesn't seem to catch the back button. Or rather, it only catches the back button if the destination is a URL that isn't handled by Flow Router. If clicking the back button goes from `mysite.com/edit` to `mysite.com`, then neither your method nor `window.onbeforeunload` catches it, unfortunately. I even tried adding this:

    FlowRouter._page.back = wrapRouting(FlowRouter._page.back);

But that didn't work either. It appears that `FlowRouter._page.back` doesn't always trigger on pressing the back button.
